parser/parse.go

```diff
@@ -260,6 +260,16 @@ func (b *Builder) AddDirTo(dir string, u *types.Universe) error {
 	return b.findTypesIn(canonicalizeImportPath(b.buildPackages[dir].ImportPath), u)
 }
 
+// DirectoryToPackage returns the package path given a dir. dir must be already
+// parsed.
+func (b *Builder) DirectoryToPackagePath(dir string) (string, error) {
```

Comment:
How about this? 

Returning Package is not going to help me, because what I need is canonicalizeImportPath(package.ImportPath).

I actually feel my original code is better. This version makes the generator call context.AddDir and context.DirectoryToPackagePath, while my version only calls context.AddDirAndReturnAddedPkgPath.